src/util/type.sml

```diff
@@ -0,0 +1,152 @@
+import "util.logging";
+
+structure Type =
+struct
+
+(*
+To properly handle all the different tokens, it can be useful to give them
+'types'. We support a simple subset of the Standard ML type language: ground
+types, variables, pairs, functions, and unary type constructors. Variables are
+denoted by a preceding quote (e.g., 'a).
+For example,
+    ('a list * int) tree -> (('a * int) -> int) -> int
+would parse to
+    Function(
+        Constructor(
+            "tree",
+            Pair(
+                Constructor("list", Var "a"),
+                Ground "int")),
+        Function(
+            Function(
+                Pair(Var "a", Ground "int"),
+                Ground "int"),
+            Ground "int"))
+Note that * and -> have equal precedence and are right associative, so
+    'a * int -> int
+parses differently to
+    ('a * int) -> int
+The first is equivalent to 'a * (int -> int). This is a deviation from the SML
+standard, but is much easier to parse.
+*)
+datatype T = Ground of string
+           | Var of string
+           | Pair of T * T
+           | Function of T * T
+           | Constructor of string * T;
+
+fun occurs x (Ground _) = false
+  | occurs x (Var y) = (x = y)
+  | occurs x (Pair (s,t)) = (occurs x s orelse occurs x t)
+  | occurs x (Function (s,t)) = (occurs x s orelse occurs x t)
+  | occurs x (Constructor (_,t)) = occurs x t;
+
+fun match (Ground s) (Ground s') = (s = s')
+(*  | match (Var x) (Var y) = true*)
+  | match (Var x) t = true (*not (occurs x t)*)
+  | match t (Var x) = true (*not (occurs x t)*)
+  | match (Pair(s,t)) (Pair(s',t')) = (match s s') andalso (match t t')
+  | match (Function(s,t)) (Function(s',t')) = (match s s') andalso (match t t')
+  | match (Constructor(s,t)) (Constructor(s',t')) = (s = s') andalso (match t t')
+  | match _ _ = false;
+
+(*A lexicographic order for types, to use in dictionaries*)
+fun compare (Ground s, Ground s') = String.compare (s,s')
+  | compare (Ground _, _) = LESS
+  | compare (_, Ground _) = GREATER
+  | compare (Var s, Var s') = String.compare (s,s')
+  | compare (Var _, _) = LESS
+  | compare (_, Var _) = GREATER
+  | compare (Pair (t,u), Pair (t',u')) = let val c = compare (t,t')
+                                         in if c = EQUAL then compare (u,u') else c
+                                         end
+  | compare (Pair _, _) = LESS
+  | compare (_, Pair _) = GREATER
+  | compare (Function (t,u), Function (t',u')) = let val c = compare (t,t')
+                                                 in if c = EQUAL then compare (u,u') else c
+                                                 end
+  | compare (Function _, _) = LESS
+  | compare (_, Function _) = GREATER
+  | compare (Constructor (s,t), Constructor (s',t')) = let val c = String.compare (s,s')
+                                                       in if c = EQUAL then compare (t,t') else c
+                                                       end
+  ;
+
+fun typeToString (Ground s) = s
```

Comment:
Please rename to `toString` for consistency.

Reply:
Done